graph/graph.go

```diff
@@ -40,6 +45,27 @@ type Graph interface {
 	Edge(u, v Node) Edge
 }
 
+// Weighted is a weighted graph.
+type Weighted interface {
+	Graph
+
+	// WeightedEdge returns the weighted edge from u to v if
+	// such an edge exists and nil otherwise. The node v must
+	// be directly reachable from u as defined by the
+	// From method.
+	WeightedEdge(u, v Node) WeightedEdge
+
+	// Weight returns the weight for the edge between
+	// x and y if Edge(x, y) returns a non-nil Edge.
+	// If x and y are the same node or there is no
+	// joining edge between the two nodes the weight
+	// value returned is implementation dependent.
+	// Weight returns true if an edge exists between
+	// x and y or if x and y have the same ID, false
```

Comment:
> ... if an edge exists between x and y or if x and y have the same ID ...

This formulation exists to support self-edges? I know simple doesn't (and shouldn't) support self-edges. Do any of the other graph implementations currently support self-edges? I'd like to play around with them in such case.